### ISSUE TITLE
URL's with # not being written correctly

### ISSUE THREAD

**Issue by dharrigan** [thymeleaf/thymeleaf#289]
Hi,

Java 1.8u5 b13.
Thymeleaf 2.1.3

I have the following defined in a template:

```
<a th:href="@{${resetTokenLink}(r=${token})}"><span th:text="${resetTokenLink}">blah</span></a>
```

I'm passing into my Context, the following:

```
context.setVariable("resetTokenLink", getDomain() + '/' + resetTokenLink);
context.setVariable("token", token);
```

getDomain() returns this:

```
https://www.domain.com
```

and resetTokenLink variable is defined as:

```
#/resetToken
```

Inside the Context (when I observe with a breakpoint), I'm seeing this:

```
resetTokenLink=https://www.domain.com/#/resetToken
token=kjsk2j3k23
```

I'm expecting this (URL) to be rendered in the template:

```
https://www.domain.com/#/resetToken?r=kjsk2j3k23
```

Instead, I'm getting this:

```
<a shape="rect" href="https://www.domain.com/?r=kjsk2j3k23#/resetToken"><span>https://www.domain.com/#/resetToken</span></a>
```

Notice how the "r" comes _before_ the "#" and the URL is malformed.

Perhaps a bug?

-=david=-

**Comment by willdavidwarren** [comment]
Isn't your desired result actually an invalid URL? I thought the "fragment" had to come very last in a URL http://en.wikipedia.org/wiki/Fragment_identifier

**Comment by dharrigan** [comment]
Hi,

Thanks for your comment. I'm not sure, since with Single Page Applications (such as AngularJS which is my target client), it appears that the format of the url:

domain/#/page?params

is valid.

It would appear the the "#" should proceed the parameters passed into a page.

Taken directly from AngularJS website, this is an example they provide:

```
http://server.com/index.html#/Chapter/1/Section/2?search=moby
```

here is another one from their site:

```
http://example.com/#/some/path?foo=bar&baz=xoxo
```

If I am wrong on this, then my apologies :-)

-=david=-

**Comment by willdavidwarren** [comment]
I'm unfamiliar with AngularJS, but I believe you. So since you're building a complete URL in the java code, why not simply output the URL using the `$` instead of the `@` and do some string concatenation. So your HTML would become:

```
<a th:href="${resetTokenLink} + 'r=' + ${token}"><span th:text="${resetTokenLink}">blah</span></a>
```

Does that work?

**Comment by dharrigan** [comment]
Hi,

Thanks again :-) Perhaps a valid workaround for now, however, this leaves this potential bug still open (and I would rather use the provided () construct as outlined in the Thymeleaf usage guide :-)).

You've been very helpful :-)

-=david=-

**Comment by willdavidwarren** [comment]
Yes, there's definitely a gap there unless there's a proper way to do this that I'm unaware of. I'm just a lowly thymeleaf user like yourself. Maybe someone like @danielfernandez would know the proper answer.

Cheers!

**Comment by danielfernandez** [comment]
Thymeleaf considers everything after a `#` in an URL base (in your case, the result of evaluating `${resetTokenLink}`) a _page fragment identifier_, and therefore a piece of information to be processed by the browser, not the server. 

Your `?r=kjsk2j3k23` parameter is not really meant for a server, but instead for your client-side library (AngularJS, in this case). But Thymeleaf does not understand this. The `(...)` syntax for specifying URL parameters is meant for _query string_ parameters (i.e. server-oriented), not _fragment_ parameters. That's why it moves the fragment to the end of the URL on purpose.

I'll leave this ticket open as a feature request therefore, so that we can think of a way to support this scenario of fragment parameters...

As a workaround, you can create your URL by appending your parameters manually, though you will have to perform URL-encoding of your parameter values yourself. For example, you can use a `|...|` _literal substitution_ like: `@{|${resetTokenLink}?r=${token}|}`

**Comment by dharrigan** [comment]
Hi Daniel,

Thank you for the update. Everything you have said makes complete sense.
I've used your suggested work-around for now :-). I guess with SPA's
becoming more popular, this type of URL pattern will become more pervalent,
but at least it's very easy to work around if required (or rather, perhaps
just adopt this as the de facto way of handling SPA fragment parameters
with Thymeleaf).

Keep up the great work on your amazingly brilliant Framework :-)

-=david=-

On 16 May 2014 23:47, Daniel Fernández notifications@github.com wrote:

> Thymeleaf considers everything after a # in an URL base (in your case,
> the result of evaluating ${resetTokenLink}) a _page fragment identifier_,
> and therefore a piece of information to be processed by the browser, not
> the server.
> 
> Your ?r=kjsk2j3k23 parameter is not really meant for a server, but
> instead for your client-side library (AngularJS, in this case). But
> Thymeleaf does not understand this. The (...) syntax for specifying URL
> parameters is meant for _query string_ parameters (i.e. server-oriented),
> not _fragment_ parameters. That's why it moves the fragment to the end of
> the URL on purpose.
> 
> I'll leave this ticket open as a feature request therefore, so that we can
> think of a way to support this scenario of fragment parameters...
> 
> As a workaround, you can create your URL by appending your parameters
> manually, though you will have to perform URL-encoding of your parameter
> values yourself. For example, you can use a |...| _literal substitution_like:
> @{|${resetTokenLink}?r=${token}|}
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thymeleaf/thymeleaf/issues/289#issuecomment-43387394
> .

## 

I prefer encrypted and signed messages. KeyID: B20A22F9
Fingerprint: 110A F423 3647 54E2 880F ADAD 1C52 85BF B20A 22F9

"It is not usually until you've built and used a version of the program
that you understand the issues well enough to get the design right." - Rob
Pike, Brian Kernighan.

No trees were harmed in the sending of this message, however, a number of
electrons were inconvenienced.

**Comment by danielfernandez** [comment]
After some research, I've decided to not create a specific syntax for parameters in URI fragment identifiers (as the ones you have in your AngularJS URLs).

The reason is, each client-side framework can do basically whatever it wants with what is specified at the URI fragment identifiers (`#...`), and the fact that AngularJS uses a query string almost equal in syntax to that of the main part of the URI can be considered a mere syntactic coincidence. The only real syntactic restriction that should be applied to fragment identifiers is that they are correctly escaped (some characters are not allowed in that part of the URI). 

So IMO there is no real point in creating thymeleaf general link syntax artifacts for these AngularJS-specific query strings.

BUT what we can offer is some generic tools that ease the work of the developer regarding the manual creation of these URLs (e.g. using literal substutitions `|...|`). So we just added a new `#uris` expression utility object (see https://github.com/thymeleaf/thymeleaf/issues/305) based on my new [Unbescape](http://www.unbescape.org) library that can be used for escaping diverse parts of the URI/URLs: paths, path segments, query parameters AND fragment identifiers.

I'm therefore closing this in favour of #305, which is already implemented and available in 2.1.4-SNAPSHOT.

**Comment by dharrigan** [comment]
Hi,

Thank you Daniel for the consideration of this issue. I understand where
you are coming from and it seems the best approach :-)

Thanks again for a great framework and tools! :-)

-=david=-

On 6 August 2014 20:30, Daniel Fernández notifications@github.com wrote:

> After some research, I've decided to not create a specific syntax for
> parameters in URI fragment identifiers (as the ones you have in your
> AngularJS URLs).
> 
> The reason is, each client-side framework can do basically whatever it
> wants with what is specified at the URI fragment identifiers (#...), and
> the fact that AngularJS uses a query string almost equal in syntax to that
> of the main part of the URI can be considered a mere syntactic coincidence.
> The only real syntactic restriction that should be applied to fragment
> identifiers is that they are correctly escaped (some characters are not
> allowed in that part of the URI).
> 
> So IMO there is no real point in creating thymeleaf general link syntax
> artifacts for these AngularJS-specific query strings.
> 
> BUT what we can offer is some generic tools that ease the work of the
> developer regarding the manual creation of these URLs (e.g. using literal
> substutitions |...|). So we just added a new #uris expression utility
> object (see #305 https://github.com/thymeleaf/thymeleaf/issues/305)
> based on my new Unbescape http://www.unbescape.org library that can be
> used for escaping diverse parts of the URI/URLs: paths, path segments,
> query parameters AND fragment identifiers.
> 
> I'm therefore closing this in favour of #305
> https://github.com/thymeleaf/thymeleaf/issues/305, which is already
> implemented and available in 2.1.4-SNAPSHOT.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/thymeleaf/thymeleaf/issues/289#issuecomment-51392662.

## 

I prefer encrypted and signed messages. KeyID: B20A22F9
Fingerprint: 110A F423 3647 54E2 880F ADAD 1C52 85BF B20A 22F9

"It is not usually until you've built and used a version of the program
that you understand the issues well enough to get the design right." - Rob
Pike, Brian Kernighan.

No trees were harmed in the sending of this message, however, a number of
electrons were inconvenienced.

**Comment by smaziano** [comment]
Just in case somebody hits this problem:
just build your url and pass it to the template:
`<a th:href="${resetUrl}">link</a>`

This should work fine.
